xxxx-ewallet-blockchain-integration.md

```diff
@@ -0,0 +1,313 @@
+2. Once the purchase is settled, a transaction is signed and sent from the eWallet to the address given by the user.
+3. The eWallet will keep an eye out on the transaction to see that it proceeds as expected, reporting the status to the user through websockets.
+
+### Transaction Flows for two eWallets
+
+For the flows below, we'll use the following use-case:
+
+```
+A end-user wants to purchase something from eWallet A using a mobile application published by (and therefore connected to) eWallet B.
+```
+
+In the cases below, both eWallets have funds deposited in Plasma: eWallet A has `ABC` tokens while eWallet B has `ZYX` tokens.
+
+#### eWallet A (Full custody) -> Ethereum address with regular transaction
+
+1. Using eWallet B mobile app, a user wants to send funds to an Ethereum address.
+2. The user submit the transaction to eWallet B, specifying how many tokens should be transferred and to which address.
+3. eWallet B submits the transaction and let user knows about its progress/finality.
+
+#### eWallet A (Full custody) -> eWallet B (Full custody) with regular transaction
+
+1. Using eWallet B mobile app, a user wants to send funds to a user of eWallet A.
+2. When submitting the transaction, the user needs to specify an eWallet's public key as well as an internal wallet address.
+3. eWallet B sends the transaction to eWallet A on Plasma with some metadata containing the details of the transaction.
+4. eWallet A receives the transaction and sends an event in the websocket channels for the given address, allowing eWallet A's user to be notified.
+
+#### eWallet A (Full custody) -> eWallet B (Full custody) with transaction request
+
+1. Using eWallet B mobile app, the user scans a QR code to pay for his purchase at eWallet A (10 `ABC`) using funds stored in eWallet B pooled wallets (`ZYX`). That QR code contains a string composed of an `ewallet_pub_key` and a transaction request ID.
+2. eWallet B mobile application calls the `transaction_request.consume` endpoint sending the content of the QR code.
+3. eWallet B detects that the received public key doesn't match its own, meaning the transaction request is hosted by another eWallet.
+4. eWallet B then makes a call to the eWallet Ethereum Name Service sending the public key received in the consumption call. A domain name will be returned by the NS.
+5. eWallet B sends a request to eWallet A to consume the request on behalf of the user. It receives a consumption ID back (with status `awaiting_funds`) and whether or not the tokens need to be converted (and if yes, to which currency) - this is defined by a conversion strategy inside the eWallet. In this case, eWallet A only wants to receive `ABC`. eWallet B connects to the websocket channel for the consumption.
+6. eWallet B requests a quote from the DEX in order to know how many `ZYX` are needed to send 10 `ABC` to eWallet A. An event is sent to eWallet A about the progress.
```

Comment:
I wonder if the quote is timely enough to make the exchange. The orderbook may have already changed since the inquiry. Should the exchange still be done or cancelled?

Maybe the user needs to set an acceptance threshold that the exchange should still happen within +/- 2% of the quoted rate?

Or maybe it's possible to setup a counterparty that's willing to accept the exchange rate risk by charging a percentage fee, similar to the Plasma's fast withdrawal counterparty? :S

Reply:
Yes this is a very good point. Usually, an exchange doesn't provide a quote for a guaranteed price. The exchange will have indicative information on the order books what the price may be and also what the last traded prices were.

When we wrote user stories, for the ideal user flow, we required a `broker` persona that could provide a quote. Note that the `broker` could be a service from the exchange, or a third party.